db/write_batch.cc

```diff
@@ -574,6 +574,13 @@ size_t WriteBatchInternal::GetFirstOffset(WriteBatch* b) {
 
 Status WriteBatchInternal::Put(WriteBatch* b, uint32_t column_family_id,
                                const Slice& key, const Slice& value) {
+  if (key.size() > size_t{port::kMaxUint32}) {
+    return Status::InvalidArgument("key is too large");
+  }
+  if (value.size() > size_t{port::kMaxUint32}) {
+    return Status::InvalidArgument("value is too large");
+  }
+
```

Comment:
These new key/value size checks along with the common lines 849-855 can be pulled into a new inline function (similar to CheckSlicePartsLength).

Reply:
I'm concerned about the performance regression of newing a Status::OK() object in normal case, so I'm careful about it. I did do it in the SliceParts path, because I don't think we care about performance there.